File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/presentation/details/TimerDetailsViewModel.java
package com.igorganapolsky.vibratingwatchapp.presentation.details;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.igorganapolsky.vibratingwatchapp.domain.repo.Repository;
import com.igorganapolsky.vibratingwatchapp.domain.model.CountData;
import com.igorganapolsky.vibratingwatchapp.domain.model.TimerModel;
import com.igorganapolsky.vibratingwatchapp.core.timer.ICountdownManager;
import com.igorganapolsky.vibratingwatchapp.core.timer.TickListener;
import com.igorganapolsky.vibratingwatchapp.util.TimerTransform;

public class TimerDetailsViewModel extends ViewModel implements TickListener {

    private final Repository repository;
    private final ICountdownManager countdownManager;

    private final MutableLiveData<CountData> activeTimerData = new MutableLiveData<>();
    private final MutableLiveData<TimerModel.State> viewStateData = new MutableLiveData<>();

    private TimerModel currentTimer;
    private int currentId = TimerModel.UNDEFINE_ID;
    private final CountData countData;

    public TimerDetailsViewModel(Repository repository, ICountdownManager countdownManager) {
        this.repository = repository;
        this.countData = CountData.getDefault();
        this.countdownManager = countdownManager;
        this.countdownManager.setTickListener(this);
    }

    LiveData<CountData> getActiveTimerData() {
        return activeTimerData;
    }

    LiveData<TimerModel.State> getViewStateData() {
        return viewStateData;
    }

    @Override
    public void onTick(Long timeLeft, int progress) {
        if (currentId == countdownManager.getActiveId()) {
            updateCountData(timeLeft, progress, true);
            activeTimerData.setValue(countData);
        }
    }

    @Override
    public void onLapEnd(Long timeLeft, int progress) {
        if (currentId == countdownManager.getActiveId()) {
            viewStateData.setValue(TimerModel.State.BEEPING);
            updateCountData(timeLeft, progress, true);
            activeTimerData.setValue(countData);
        }
    }

    @Override
    public void onFinish(Long timeLeft, int progress, boolean isStop) {
        if (currentId == countdownManager.getActiveId()) {
            viewStateData.setValue(TimerModel.State.FINISHED);
            updateCountData(timeLeft, progress, isStop);
            activeTimerData.setValue(countData);
        }
    }

    private void updateCountData(Long timeLeft, int progress, boolean isAnimationNeeded) {
        countData.setCurrentTime(TimerTransform.millisToString(timeLeft));
        countData.setCurrentProgress(progress);
    }

    void prepareData(int currentId) {
        boolean isActive = defineCurrentTimer(currentId);
        long timeToSetup = prepareTime(isActive);
        int progress = prepareProgress(isActive);
        updateState(timeToSetup, progress);
    }

    private boolean defineCurrentTimer(int currentId) {
        this.currentId = currentId;
        int activeId = countdownManager.getActiveId();
        boolean isActive = this.currentId == activeId && countdownManager.isActive();

        if (isActive) {
            currentTimer = countdownManager.getActive();
        } else {
            currentTimer = repository.getTimerById(this.currentId);
        }
        return isActive;
    }

    void checkUpdates() {
        currentTimer = repository.getTimerById(currentId);
        countdownManager.setupTimer(currentTimer);
        long timeToSetup = prepareTime(true);
        updateState(timeToSetup, 100);
    }

    private void updateState(long timeToSetup, int progress) {
        viewStateData.setValue(currentTimer.getState());
        activeTimerData.setValue(new CountData(
            TimerTransform.millisToString(timeToSetup),
            progress,
            false));
    }

    private long prepareTime(boolean isActive) {
        if (isActive) {
            return countdownManager.getActiveTimeLeft();
        } else {
            return TimerTransform.timeToMillis(
                currentTimer.getHours(),
                currentTimer.getMinutes(),
                currentTimer.getSeconds());
        }
    }

    private int prepareProgress(boolean isActive) {
        if (isActive) {
            return countdownManager.getActiveProgress();
        } else {
            return 100;
        }
    }

    void onStart() {
        viewStateData.setValue(TimerModel.State.RUNNING);
        int currentActiveTimerId = countdownManager.getActiveId();

        if (currentTimer.getId() == currentActiveTimerId) {
            countdownManager.onStart();
        } else {
            if (currentActiveTimerId != TimerModel.UNDEFINE_ID) {
                countdownManager.onStop();
            }
            countdownManager.setupTimer(currentTimer);
            countdownManager.onStart();
        }
    }

    void onPause() {
        viewStateData.setValue(TimerModel.State.PAUSED);
        countdownManager.onPause();
    }

    void onStop() {
        if (currentTimer.getId() == countdownManager.getActiveId()) {
            viewStateData.setValue(TimerModel.State.FINISHED);
            countdownManager.onStop();
        }
    }

    void onRestart() {
        if (currentTimer.getId() == countdownManager.getActiveId() &&
            currentTimer.getState() != TimerModel.State.FINISHED) {

            viewStateData.setValue(TimerModel.State.RUNNING);
            countdownManager.onRestart();
        }
    }

    void onNextLap() {
        if (currentTimer.getId() == countdownManager.getActiveId()) {
            boolean isNextLapStarted = countdownManager.onNextLap();
            TimerModel.State newState = isNextLapStarted ? TimerModel.State.RUNNING : TimerModel.State.FINISHED;
            viewStateData.setValue(newState);

        }
    }

    public void deleteTimer() {
        repository.deleteTimer(currentTimer.getId());
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        countdownManager.setTickListener(null);
    }
}

File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/presentation/details_screen/TimerDetailsViewModel.kt
package com.igorganapolsky.vibratingwatchapp.presentation.details_screen

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.igorganapolsky.vibratingwatchapp.core.ICountdownController
import com.igorganapolsky.vibratingwatchapp.core.ITickListener
import com.igorganapolsky.vibratingwatchapp.core.TimerTransform
import com.igorganapolsky.vibratingwatchapp.data.CountModel
import com.igorganapolsky.vibratingwatchapp.data.IRepository
import com.igorganapolsky.vibratingwatchapp.domain.model.TimerModel

class TimerDetailsViewModel(
    private val repository: IRepository,
    private val countdownController: ICountdownController
) : ViewModel(), ITickListener {

    internal val activeTimerData = MutableLiveData<CountModel>()
    val viewStateData = MutableLiveData<TimerModel.State>()

    private lateinit var currentTimer: TimerModel
    private var currentId = TimerModel.UNDEFINE_ID
    private val countData: CountModel = CountModel.default

    init {
        this.countdownController.setTickListener(this)
    }

    internal fun getActiveTimerData(): LiveData<CountModel> {
        return activeTimerData
    }

    internal fun getViewStateData(): LiveData<TimerModel.State> {
        return viewStateData
    }

    override fun onTick(timeLeft: Long?, progress: Int) {
        if (currentId == countdownController.activeId) {
            updateCountData(timeLeft, progress, true)
            activeTimerData.value = countData
        }
    }

    override fun onLapEnd(timeLeft: Long?, progress: Int) {
        if (currentId == countdownController.activeId) {
            viewStateData.value = TimerModel.State.BEEPING
            updateCountData(timeLeft, progress, true)
            activeTimerData.value = countData
        }
    }

    override fun onFinish(timeLeft: Long?, progress: Int, isStop: Boolean) {
        if (currentId == countdownController.activeId) {
            viewStateData.value = TimerModel.State.FINISHED
            updateCountData(timeLeft, progress, isStop)
            activeTimerData.value = countData
        }
    }

    private fun updateCountData(timeLeft: Long?, progress: Int, isAnimationNeeded: Boolean) {
        countData.currentTime = TimerTransform.millisToString(timeLeft!!)
        countData.currentProgress = progress
    }

    internal fun prepareData(currentId: Int) {
        val isActive = defineCurrentTimer(currentId)
        val timeToSetup = prepareTime(isActive)
        val progress = prepareProgress(isActive)
        updateState(timeToSetup, progress)
    }

    private fun defineCurrentTimer(currentId: Int): Boolean {
        this.currentId = currentId
        val activeId = countdownController.activeId
        val isActive = this.currentId == activeId && countdownController.isActive

        currentTimer = if (isActive) {
            countdownController.activeTimerModel
        } else {
            repository.getTimerById(this.currentId)
        }
        return isActive
    }

    internal fun checkUpdates() {
        currentTimer = repository.getTimerById(currentId)
        countdownController.setupTimer(currentTimer)
        val timeToSetup = prepareTime(true)
        updateState(timeToSetup, 100)
    }

    private fun updateState(timeToSetup: Long, progress: Int) {
        viewStateData.value = currentTimer.state
        activeTimerData.value = CountModel(
            TimerTransform.millisToString(timeToSetup),
            progress,
            false
        )
    }

    private fun prepareTime(isActive: Boolean): Long {
        return if (isActive) {
            countdownController.activeTimeLeft
        } else {
            TimerTransform.timeToMillis(
                currentTimer.hours,
                currentTimer.minutes,
                currentTimer.seconds
            )
        }
    }

    private fun prepareProgress(isActive: Boolean): Int {
        return if (isActive) {
            countdownController.activeProgress
        } else {
            100
        }
    }

    internal fun onStart() {
        viewStateData.value = TimerModel.State.RUNNING
        val currentActiveTimerId = countdownController.activeId

        if (currentTimer.id == currentActiveTimerId) {
            countdownController.onStart()
        } else {
            if (currentActiveTimerId != TimerModel.UNDEFINE_ID) {
                countdownController.onStop()
            }
            countdownController.setupTimer(currentTimer)
            countdownController.onStart()
        }
    }

    internal fun onPause() {
        viewStateData.value = TimerModel.State.PAUSED
        countdownController.onPause()
    }

    internal fun onStop() {
        if (currentTimer.id == countdownController.activeId) {
            viewStateData.value = TimerModel.State.FINISHED
            countdownController.onStop()
        }
    }

    internal fun onRestart() {
        if (currentTimer.id == countdownController.activeId && currentTimer.state !== TimerModel.State.FINISHED) {

            viewStateData.value = TimerModel.State.RUNNING
            countdownController.onRestart()
        }
    }

    internal fun onNextLap() {
        if (currentTimer.id == countdownController.activeId) {
            val isNextLapStarted = countdownController.onNextLap()
            val newState =
                if (isNextLapStarted) TimerModel.State.RUNNING else TimerModel.State.FINISHED
            viewStateData.value = newState

        }
    }

    fun deleteTimer() {
        repository.deleteTimer(currentTimer.id)
    }

    override fun onCleared() {
        super.onCleared()
        countdownController.setTickListener(null)
    }
}
